### PR TITLE
Heuristics for variable naming

### PR DESCRIPTION
**Pull request by GeraldineGalindo**
Hi! This small PR proposes a new mode of using methods and arguments information (obtained via reflection) for naming variables in EvoSuite-generated tests. 

Currently, the variable naming is based on the variable type with an index number indicating the number of repetitions the name presents. To improve readability this change proposes using little heuristics on the `TestCodeVisitor` class to gather code information and rename variables.  Consider the following examples of `test3`, The first one uses `linkedList0` and `linkedList1` for naming two variables sent as arguments of the `BookStore` class.
```
  public void test3()  throws Throwable  {
      LinkedList<Book> linkedList0 = new LinkedList<Book>();
      LinkedList<Author> linkedList1 = new LinkedList<Author>();
      BookStore bookStore0 = new BookStore(linkedList0, linkedList1);
     ...
  }
```

In this case, we propose to use information about the names of the arguments in the `BookStore` declaration to give more meaningful names (related to the class under test) as seen in the following version of `test3`:

```
public void test3()  throws Throwable  {
      LinkedList<Book> books = new LinkedList<Book>();
      LinkedList<Author> authors = new LinkedList<Author>();
      BookStore bookStore = new BookStore(books, authors);
     ...
  }
```
Our little heuristics implemented in this PR include using method argument names, and method names (+ a little modification on the first words of the suggested name). Please let us know your thoughts/suggestions about it!

To test the additions in this PR, change the attribute `VariableNamingStrategy` in the `Properties` class to: 
```
@Parameter(key = "variable_naming_strategy", group = "Output", description = "What strategy to use to name variables for tests")
public static VariableNamingStrategy VARIABLE_NAMING_STRATEGY = VariableNamingStrategy.HEURISTICS_BASED;
```

Then, compile to obtain the `jar` file. For generating tests, you must compile the project under test with the option `-parameters` configured for `javac`, this is used to obtain the methods argument names by reflection. You can configure this option in your IDE (Settings > Build, Execution, Deployment > Compiler> Java Compiler > Additional compile line parameters) in IntelliJ)or the `pom.xml` file in Maven.

### CODE REVIEW

**Comment by jose** [comment]
We might need to discuss the pros/cons of pull requests #451 and #441 before going ahead.  @GeraldineGalindo any chance you could summarise the pros/cons of this pull request vs. #441?

**Comment by GeraldineGalindo** [comment]
> We might need to discuss the pros/cons of pull requests #451 and #441 before going ahead. @GeraldineGalindo any chance you could summarise the pros/cons of this pull request vs. #441?

Thanks for the comment! Sure: 

As I reviewed, #441 proposes a good structure for managing the responsibilities in the `TestCodeVisitor` class. The use of a strategy pattern opens more possibilities, heuristics proposed in #451 can be implemented with this structure, for less coupling. 

I think that the main difference between both PRs is the focus of each one, as I understand #441 proposes the base structure for using different naming strategies. In contrast, #451 proposes two heuristics besides the type-based one using information gathered in visiting phase (which is why this PR introduces changes to the `TestCodeVisitor` class), therefore, I do not think that the PRs are comparable, instead, they can be complementary. 

If more strategies could be crafted for variable naming in EvoSuite, a structure defined in #441 is needed to avoid the `TestCodeVisitor` class becoming a large class. However, this is designed to only use the information of the variable types obtained from the `VariableReference` class. #451 uses dictionaries for gathering and crafting variable names based on reflection information retrieval (so, adapting #451 to the structure in #441 would need some changes). If possible, I'd like to make a new PR adapting the heuristics proposed with the strategy pattern structure. What do you think about this option?

**Comment by GeraldineGalindo** [comment]
@jose, thanks again for the comments, we reviewed PR #441 again and did a little test to implement our naming suggestion with the proposed structure successfully. So, we kindly inquire about the status of #441, and if its refactoring would be included in EvoSuite, we'd like to upload a PR with the implementation using this new structure, and if it helps we'd be glad to meet and discuss the PR. Many thanks for considering my request.

**Comment by jose** [comment]
Thanks @GeraldineGalindo for your feedback.  I'm going to merge [#441](https://github.com/EvoSuite/evosuite/pull/441).  It would be great if could adapt/update [#451](https://github.com/EvoSuite/evosuite/pull/451) to use the new structure proposed in [#441](https://github.com/EvoSuite/evosuite/pull/441).

**Comment by GeraldineGalindo** [comment]
Thanks @jose for the answer! I am uploading the update of this commit to the new structure and I'll be attentive for comments or reviews on the proposal.

File: client/src/main/java/org/evosuite/Properties.java [hunk 1]
```diff
@@ -1242,11 +1244,11 @@ public enum TestNamingStrategy {
     public static TestNamingStrategy TEST_NAMING_STRATEGY = TestNamingStrategy.NUMBERED;
 
     public enum VariableNamingStrategy {
-        TYPE_BASED
+        TYPE_BASED, HEURISTICS_BASED
     }
 
     @Parameter(key = "variable_naming_strategy", group = "Output", description = "What strategy to use to derive names for variables")
-    public static VariableNamingStrategy VARIABLE_NAMING_STRATEGY = VariableNamingStrategy.TYPE_BASED;
+    public static VariableNamingStrategy VARIABLE_NAMING_STRATEGY = VariableNamingStrategy.HEURISTICS_BASED;
```
Comment (jose) [inline client/src/main/java/org/evosuite/Properties.java:1251]: I'm inclined to keep the variable naming strategy as `TYPE_BASED`, by default, for now.

File: client/src/main/java/org/evosuite/testcase/variable/name/TypeBasedVariableNameStrategy.java
Comment (jose) [inline client/src/main/java/org/evosuite/testcase/variable/name/TypeBasedVariableNameStrategy.java:1]: The changes you made to this file were just refactors, right?  Or is this here anything new I'm not currently seeing?

**Review by jose — CHANGES_REQUESTED** [review]

File: client/src/main/java/org/evosuite/Properties.java (hunk 1 above)
Comment (GeraldineGalindo) [reply client/src/main/java/org/evosuite/Properties.java:1251]: Yes, definitely, I'm sorry, I thought I uploaded it with `TYPE_BASED` by default, the heuristics proposed are an option. I'll correct it right away.

File: client/src/main/java/org/evosuite/testcase/variable/name/TypeBasedVariableNameStrategy.java
Comment (GeraldineGalindo) [reply client/src/main/java/org/evosuite/testcase/variable/name/TypeBasedVariableNameStrategy.java:1]: Yes, the intention is to split the the process of getting the type name and adding an index to it. A change on the heuristics proposal is to not use indexes when there is no repetition between variable names. However, the type-based renaming always adds an index; as the heuristics use the type name as base suggestion, I refactored this class to get the type name before using index.

File: client/src/main/java/org/evosuite/Properties.java (hunk 1 above)
Comment (jose) [reply client/src/main/java/org/evosuite/Properties.java]: Thanks.

File: client/src/main/java/org/evosuite/testcase/variable/name/TypeBasedVariableNameStrategy.java
Comment (jose) [reply client/src/main/java/org/evosuite/testcase/variable/name/TypeBasedVariableNameStrategy.java:1]: Got it, thanks.

**Review by jose — APPROVED** [review]

**Comment by jose** [comment]
Thanks @GeraldineGalindo.

**Comment by GeraldineGalindo** [comment]
Thanks @jose for reviewing and for your time :)

**Pull request merged**